### PR TITLE
decouple kwargs performance improvement from API changes

### PR DESCRIPTION
**Pull request by vtjnash**
This undoes the breaking API change introduced by #24795 (fixing #25177), while keeping the performance / inference improvement. This lets us develop the kwargs API independent of Core.NT, for example, to provide proper deprecations for future changes, specific error messages, and work towards possible future development of NT (for example, implementing it as just another normal type in Julia, rather than requiring a special cases in the runtime).

### CODE REVIEW

File: base/coreimg.jl [hunk 1]
```diff
@@ -57,6 +57,7 @@ include("reduce.jl")
 include("bitarray.jl")
 include("bitset.jl")
 include("abstractdict.jl")
+include("iterators.jl")
```
Comment (JeffBezanson) [inline base/coreimg.jl]: Maybe we should move `IndexValue` to its own file?

**Comment by JeffBezanson** [comment]
I think this is fine, for the sake of fixing #25177. That covers the most noticeable API change, but there are still other breaking API changes --- kwargs not being a vector --- but I think those are fine and should stay. I just want to be clear that the performance and API changes are not completely de-coupled.

File: base/coreimg.jl (hunk 1 above)
Comment (vtjnash) [reply base/coreimg.jl]: We could, but seemed useful enough to be able to call pairs (required Zip) and generator Filter and we may want to use the Product iterator for Cartesian arrays, so I’ve wanted this file before anyways.

**Comment by bramtayl** [comment]
I've updated https://github.com/bramtayl/Keys.jl master for 0.7 to take advantage of dot overloading and constant progogation (it also requires the master version of RecurUnroll). It provides a named tuple implementation that

- Is fully type-stable
- Requires no generated functions
- Requires no compiler modifications
- Requires no Vals
- Allows for constant propagation (unlike the base implementation of named tuples). This seems essential making keyword arguments useful.

**Comment by vtjnash** [comment]
That seems interesting. What do you mean by "allows constant propagation"?

**Comment by bramtayl** [comment]
Hmm, wait, maybe I was confused about that. 

I was looking at

```
test() = KeyedTuple(; a = 1, b = 2.0).a
@code_warntype test()
# return 1
```

But this seems to work with named tuples as well:

```
test() = (a = 1, b = 2.0).a
@code_warntype test()
# return 1
```

I thought I had read somewhere that constant propagation doesn't survive named tuples?

**Comment by bramtayl** [comment]
https://github.com/JuliaLang/julia/issues/25188#issuecomment-354291498

**Comment by vtjnash** [comment]
> I thought I had read somewhere that constant propagation doesn't survive named tuples?

Here's an example of that:
```julia
test() = (a = 1, b = 2.0).a + 1
```

**Comment by bramtayl** [comment]
Hmm, I see basically identical generated code to 

```
test() = KeyedTuple(; a = 1, b = 2.0).a + 1
```

It seems like it propagates up to
```
SSAValue(67) = (Base.add_int)(1, 1)::Int64
return SSAValue(67)
```

File: src/julia-syntax.scm [hunk 2]
```diff
@@ -512,13 +512,13 @@
...
              ,(if (any kwarg? pargl) (gensy) UNUSED)
-             (call (core kwftype) ,ftype)) (:: ,kw (core NamedTuple)) ,@pargl ,@vararg)
+             (call (core kwftype) ,ftype)) ,kw ,@pargl ,@vararg)
           `(block
             ,(scopenest
               keynames
               (map (lambda (v dflt)
                      (let* ((k     (decl-var v))
-                            (rval0 `(call (top getfield) ,kw (quote ,k)))
+                            (rval0 `(call (top getindex) ,kw (inert ,k)))
```
Comment (JeffBezanson) [inline src/julia-syntax.scm]: Will constant prop work here? I thought it was only enabled for `getproperty` currently.

**Pull request merged**